app/src/main/java/com/epam/esm/model/dao/impl/SqlGiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,222 @@
+                addNestedTags(patch);
+            }
+            jdbcTemplate.update(UPDATE_GIFT_CERTIFICATE_QUERY,
+                    id,
+                    giftCertificate.getName(),
+                    giftCertificate.getDescription(),
+                    giftCertificate.getPrice(),
+                    giftCertificate.getDuration(),
+                    giftCertificate.getCreateDate() != null
+                            ? Timestamp.valueOf(giftCertificate.getCreateDate())
+                            : null,
+                    giftCertificate.getLastUpdateDate() != null
+                            ? Timestamp.valueOf(giftCertificate.getLastUpdateDate())
+                            : null,
+                    id);
+        } else {
+            throw new DaoException("Gift certificate with such id is not exist");
+        }
+    }
+
+    @Override
+    public void delete(long id) throws DaoException {
+        jdbcTemplate.update(DELETE_QUERY, id);
+    }
+
+    public List<GiftCertificate> findByTagId(long id) {
+        return jdbcTemplate.query(FIND_BY_TAG_ID_QUERY, new GiftCertificateRowMapper(), id);
+    }
+
+    @Override
+    public List<GiftCertificate> findByTagName(String tagName) throws DaoException {
+        return findByTagName(tagName, new QueryCustomizer());
+    }
+
+    @Override
+    public List<GiftCertificate> findByTagName(String tagName, QueryCustomizer queryCustomizer) throws DaoException {
+        List<GiftCertificate> giftCertificates = jdbcTemplate.query(queryCustomizer.prepareQuery(FIND_BY_TAG_NAME_QUERY), new GiftCertificateRowMapper(), tagName);
+        for (GiftCertificate giftCertificate : giftCertificates) {
+            List<Tag> tags = tagDao.findByGiftCertificateId(giftCertificate.getId());
+            giftCertificate.setTags(tags);
+        }
+        return giftCertificates;
+    }
+
+    private void addNestedTags(GiftCertificate giftCertificate) throws DaoException {
+        if (giftCertificate.getTags() != null) {
+            for (Tag tag : giftCertificate.getTags()) {
+                try {
+                    // Adding tag into database if id is passed and such tag is not exist
+                    if (tag.getId() == null) {
+                        long generatedTagId = tagDao.add(tag);
+                        tag.setId(generatedTagId);
+                    } else {
+                        if (tagDao.findById(tag.getId()).isEmpty()) {
+                            long generatedTagId = tagDao.add(tag);
+                            tag.setId(generatedTagId);
+                        }
+                    }
+                    jdbcTemplate.update(INSERT_INTO_REF_TABLE_QUERY, giftCertificate.getId(), tag.getId());
+                } catch (DaoException e) {
+                    throw new DaoException("Error while adding tag", e);
+                }
+            }
+        }
+    }
+
+    private static class GiftCertificateRowMapper implements RowMapper<GiftCertificate> {
```

Comment:
check spring's ResultSetExtractor

Reply:
Should i rewrite some methods using ResultSetExtractor or just read about it?